src/startApp.ts

```diff
@@ -88,13 +99,40 @@ async function loadEntryPointFile(functionAppDirectory: string): Promise<void> {
             }
         } catch (err) {
             const error = ensureErrorType(err);
-            worker.log({
-                message: `Worker was unable to load entry point "${currentFile ? currentFile : entryPointPattern}": ${
-                    error.message
-                }`,
-                level: LogLevel.Warning,
-                logCategory: LogCategory.System,
-            });
+            const newMessage = `Worker was unable to load entry point "${currentFile || entryPointPattern}": ${
+                error.message
+            }`;
+
+            if (shouldBlockOnEntryPointError()) {
+                // This is blocking. The event handler will respond saying it failed
+                error.message = newMessage;
+                throw error;
+            } else {
+                // This is not blocking. We directly log the error, but the event handler will respond saying it succeeded
+                worker.log({
+                    message: newMessage,
+                    level: LogLevel.Error,
+                    logCategory: LogCategory.System,
+                });
+            }
+        }
+    }
+}
+
+function shouldBlockOnEntryPointError(): boolean {
+    if (isNode20Plus()) {
+        // Starting with Node 20, this will always be blocking
+        // https://github.com/Azure/azure-functions-nodejs-worker/issues/697
+        return true;
+    } else {
+        const key = 'FUNCTIONS_NODE_BLOCK_ON_ENTRY_POINT_ERROR';
+        if (isDefined(process.env[key])) {
+            return isEnvironmentVariableSet(process.env[key]);
+        } else {
+            // We think this should be a blocking error by default, but v3 can't do that for backwards compatibility reasons
+            // https://github.com/Azure/azure-functions-nodejs-worker/issues/630
+            const model = nonNullProp(worker.app, 'programmingModel');
+            return !(model.name === '@azure/functions' && model.version.startsWith('3.'));
```

Comment:
What happens in the case the model isn't set yet?

Reply:
That _should_ be impossible - the default programming model is set before worker init

Reply:
I thought one of the arguments for having the app setting was for cases where the error happens before the default programming model is set?

Reply:
The default is v3. I think you're thinking of the case where an error happens before the new v4 model is set

Reply:
Oh makes sense now